shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+-- toValBST :: Value crypto -> ValueBSType
+-- toValBST (Value v) = ValueBSType $ fromList $ fmap (\(cid, tkns) -> (pack $ show cid, tkns)) (toList v)
+
+-- | compact representation of Value
+data CompactValue crypto = AdaOnly Coin | MixValue (Value crypto)
+  deriving (Show, Eq, Generic)
+
+instance NoUnexpectedThunks (CompactValue crypto)
+
+
+-- get the quantities of the tokens of a value term
+getQs :: Value crypto -> [Quantity]
+getQs (Value v) = fmap snd (concat $ fmap toList (elems v))
+
+-- zero value
+zeroV :: Value crypto
+zeroV = Value empty
+
+-- | make a value out of a coin
+coinToValue :: Coin -> Value crypto
+coinToValue (Coin c) = Value $ singleton adaID (singleton adaToken (Quantity c))
+
+-- | make a value out of a coin
+getAdaAmount :: Value crypto -> Coin
+getAdaAmount (Value v) = Coin $ c
+  where
+    Quantity c = foldl (+) (Quantity 0) (getQs $ Value $ filterWithKey (\k _ -> k == adaID) v)
+
+-- | policy ID of Ada
+adaID :: PolicyID crypto
+adaID = PolicyID $ ScriptHash $ coerce ("" :: Hash ShortHash (Script crypto))
+
+-- | asset ID of Ada
+adaToken :: AssetID
+adaToken = AssetID $ pack "Ada"
+
+valueToCompactValue :: Value crypto -> CompactValue crypto
+valueToCompactValue vl
+  | keys v == [adaID] = AdaOnly $ getAdaAmount vl
+  | otherwise         = MixValue $ uniqueAdaToken vl
+    where
+      (Value v) = vl
+
+-- | make a Value term where the only token name within the ada asset group is adaToken
+-- no 0-valued
+uniqueAdaToken :: Value crypto -> Value crypto
+uniqueAdaToken (Value v) = Value $ insert adaID
+  (singleton adaToken $ foldl (+) 0
+  (fmap snd $ concat $ fmap toList (elems (filterWithKey (\k _ -> k == adaID) v))) )
+  (removeZeros $ filterWithKey (\k _ -> k /= adaID) v)
+  where
+    removeZeros vv = fmap (filterWithKey (\_ q -> q /= (Quantity 0))) vv
+
+compactValueToValue :: CompactValue crypto -> Value crypto
+compactValueToValue (AdaOnly c)  = coinToValue c
+compactValueToValue (MixValue v) = v
+
+instance Eq (Value crypto) where
+    {-# INLINABLE (==) #-}
+    (==) = eq
+
+-- TODO these instances
+-- No 'Ord Value' instance since 'Value' is only a partial order, so 'compare' can't
+-- do the right thing in some cases.
+-- Values are compared in a lexicographical way
+-- this is needed to define an instance of Relation
+-- use Map Ord instance
+instance Ord (Value crypto) where
+   compare (Value v1) (Value v2) = compare v1 v2
+
+-- TODO this should return the same result as for Value
+instance Ord (CompactValue crypto) where
+   compare v1 v2 = compare (compactValueToValue v1) (compactValueToValue v2)
+
+instance Semigroup (Value crypto) where
+    {-# INLINABLE (<>) #-}
+    (<>) = addv
+
+instance Monoid (Value crypto) where
+    {-# INLINABLE mempty #-}
+    mempty  = zeroV
+    mappend = (<>)
+
+-- instances for CompactValue
+instance Semigroup (CompactValue crypto) where
+    {-# INLINABLE (<>) #-}
+    (<>) v1 v2 = valueToCompactValue $ addv (compactValueToValue v1) (compactValueToValue v2)
```

Comment:
This looks quite inefficient, but it probably doesn't matter.

Reply:
i am not sure. I will see about making it more efficient if it's decided to be a problem